storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1746 @@ buf_block_t *fsp_get_latched_xdes_page(
+  we store the child page number. This is stored
+  in the form of <child_page_no, parent_page_no + parent_offset> */
+  std::unordered_map<uint32_t, uint64_t> m_parent_pages;
+
+  dict_index_t &m_index;
+
+  buf_block_t *m_root;
+  /** Iterate through the page and map the child_page_no
+  with the parent page and their associate offset
+  in m_parent_pages
+  @param block  block to be traversed */
+  dberr_t store_child_pages(buf_block_t *block) noexcept;
+
+  /** Get the first block for the given level
+  @param level       level
+  @param mtr         mini-transaction
+  @param cur_page_no first page number for the given level
+  @return error code or DB_SUCCESS */
+  dberr_t get_level_block(uint16_t level, mtr_t *mtr,
+                          uint32_t *cur_page_no) noexcept;
+
+  /** Defragment the level of the index
+  @param level          level to be defragmented
+  @param mtr            mini-transaction
+  @param space_defrag   space defragmenter information
+                        and also responsible for allocating new
+                        segment or page from tablespace
+  @return error code or DB_SUCCESS */
+  dberr_t defragment_level(uint16_t level, mtr_t *mtr,
+                           SpaceDefragmenter *space_defrag) noexcept;
+
+public:
+  IndexDefragmenter(dict_index_t &index): m_index(index) {}
+
+  /** Defragment the index with the help of space defragmenter.
+  1) Iterate through each level of the index
+  2) Find out what are the pages/segment
+  to be modified for the index.
+  3) Allocate the page from the new segment/extent
+  4) Copy the to be changed page content to new page
+  5) Change the associative pages in the tree with
+  new page(left, right, parent block)
+  6) Do step (4), (5) within single mini-transaction
+  and commit the mini-transaction
+  @return error code or DB_SUCCESS */
+  dberr_t defragment(SpaceDefragmenter *space_defrag) noexcept;
+};
+
+class SpaceDefragmenter final
+{
+  /** Store the extent information in the tablespace <extent, state>*/
+  std::map<uint32_t, uint32_t> m_extent_info;
+  /** Map of last used extent with early unused extent within
+  the tablespace */
+  std::map<uint32_t, uint32_t> m_extent_map;
+
+  /** Collect the extent information from tablespace */
+  dberr_t extract_extent_state() noexcept
+  {
+    mtr_t mtr;
+    dberr_t err= DB_SUCCESS;
+    uint32_t last_descr_page_no= 0;
+    fil_space_t *space= fil_system.sys_space;
+    mtr.start();
+    mtr.x_lock_space(space);
+    buf_block_t *last_descr= buf_page_get_gen(page_id_t{space->id, 0}, 0,
+                                              RW_S_LATCH, nullptr,
+                                              BUF_GET_POSSIBLY_FREED, &mtr,
+                                              &err);
+    if (!last_descr)
+    {
+func_exit:
+      mtr.commit();
+      return err;
+    }
+
+    for (uint32_t xdes_n= 0; xdes_n < space->free_limit;
+         xdes_n+= FSP_EXTENT_SIZE)
```

Comment:
The rather complex expression `FSP_EXTENT_SIZE` is being evaluated on each iteration.